# Week 1 : Day 4 PM // Pandas Basic 2

## NBA dataset analysis

In [1]:
import numpy as np
import pandas as pd

In [ ]:
#ngeload data set

#skema 1 = local-path
# pd.read_csv('filename')
# pd.read_csv('path/filename')
# pd.read_csb('../filename')
# pd.read_csv('../../filename')

#skema 2 = separator berbeda (tanda pemisah)
#contoh data1;data2;data3
#pd.read_csv('resource/ref.csv', delimiter= ';' )

#skema 3 = link/web

## untuk file selain excel read nya menggunakan read_csv

#### skema 1

In [2]:
pd.read_csv('resource/Student_Marks.csv')

number_courses  time_study   Marks
0                3       4.508  19.202
1                4       0.096   7.734
2                4       3.133  13.811
3                6       7.909  53.018
4                8       7.811  55.299
..             ...         ...     ...
95               6       3.561  19.128
96               3       0.301   5.609
97               4       7.163  41.444
98               7       0.309  12.027
99               3       6.335  32.357

[100 rows x 3 columns]

#### skema 2

In [7]:
pd.read_csv('resource/ref.csv', delimiter= ';' )

ch1  ch2  ch3
0   10    1  100
1   20    2  200
2   30    3  300
3   40    4  400
4   50    5  500

#### Skema 3

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')

##### to show how much data nba contains:

RangeIndex(start=0, stop=126314, step=1)

In [13]:
#berisi berapa banyak baris
len(df)

126314

In [9]:
#jumlah baris dan kolom
#1236314 baris dan 23 kolom

df.shape

(126314, 23)

In [11]:
df.shape[0]

126314

In [15]:
df.shape[1]

23

In [17]:
#df.head()
#preview 5 data teratas secara default
#df.head(2)
#preview 2 data teratas

df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [18]:
#set display colums agar semua terdisplay
pd.set_option('display.max_columns', None)

In [20]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

In [22]:
#preview 5 data terbawah
#df.tail()

df.tail()

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126310         102            1     GSW   Warriors  104  1809.9791  1813.6349   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   
126313         103            1     GSW   Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309  60.309792    GSW   Warriors      103  1790.9591  1809.9791   
126310  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391   
126311  60.010067    GSW   Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW   Warriors      105  1813.6349  1822.2881   
126313  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
126309             H           L  0.546572   NaN  
126310             H           W  0.765565   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  
126313             A           W  0.518550   NaN

## Getting to Know Your Data

### Data overview

In [24]:
df.info()

#Non-Null Count = missing values
#Dtype = type data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

### Show basic statistic

In [27]:
#df.describe()
#df.describe(),T < transpose
df.describe().T

count          mean           std          min           25%  \
gameorder    126314.0  31579.000000  18231.927643     1.000000  15790.000000   
_iscopy      126314.0      0.500000      0.500002     0.000000      0.000000   
year_id      126314.0   1988.200374     17.582309  1947.000000   1975.000000   
seasongame   126314.0     43.533733     25.375178     1.000000     22.000000   
is_playoffs  126314.0      0.063857      0.244499     0.000000      0.000000   
pts          126314.0    102.729982     14.814845     0.000000     93.000000   
elo_i        126314.0   1495.236055    112.139945  1091.644500   1417.237975   
elo_n        126314.0   1495.236055    112.461687  1085.774400   1416.994900   
win_equiv    126314.0     41.707889     10.627332    10.152501     34.103035   
opp_pts      126314.0    102.729982     14.814845     0.000000     93.000000   
opp_elo_i    126314.0   1495.236055    112.139945  1091.644500   1417.237975   
opp_elo_n    126314.0   1495.236055    112.461687  1085.774400   1416.994900   
forecast     126314.0      0.500000      0.215252     0.020447      0.327989   

                      50%           75%           max  
gameorder    31579.000000  47368.000000  63157.000000  
_iscopy          0.500000      1.000000      1.000000  
year_id       1990.000000   2003.000000   2015.000000  
seasongame      43.000000     65.000000    108.000000  
is_playoffs      0.000000      0.000000      1.000000  
pts            103.000000    112.000000    186.000000  
elo_i         1500.945550   1576.060000   1853.104500  
elo_n         1500.954400   1576.291625   1853.104500  
win_equiv       42.113357     49.635328     71.112038  
opp_pts        103.000000    112.000000    186.000000  
opp_elo_i     1500.945550   1576.060000   1853.104500  
opp_elo_n     1500.954400   1576.291625   1853.104500  
forecast         0.500000      0.672011      0.979553

### Data Exploration

In [28]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

In [33]:
df['game_result'].unique()

array(['L', 'W'], dtype=object)

In [34]:
df['game_result'].value_counts()

L    63157
W    63157
Name: game_result, dtype: int64

In [39]:
df['fran_id'].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

In [43]:
df.sort_values(by='pts',ascending=True)[['year_id','fran_id','pts',]]

year_id  fran_id  pts
26684     1973  Nuggets    0
26685     1973  Squires    2
3190      1951   Lakers   18
3191      1951  Pistons   19
4         1947  Falcons   33
...        ...      ...  ...
8974      1959  Celtics  173
32881     1975    Sails  176
21416     1970   Pacers  177
50095     1984  Nuggets  184
50094     1984  Pistons  186

[126314 rows x 3 columns]

In [46]:
df.loc[df["fran_id"] == "Knicks",:]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
1               1  194611010TRH   NBA        1     1947   11/1/1946   
3               2  194611020CHS   NBA        1     1947   11/2/1946   
23             12  194611070STB   NBA        1     1947   11/7/1946   
37             19  194611110NYK   NBA        0     1947  11/11/1946   
54             28  194611160NYK   NBA        0     1947  11/16/1946   
...           ...           ...   ...      ...      ...         ...   
126025      63013  201504080NYK   NBA        0     2015    4/8/2015   
126056      63029  201504100NYK   NBA        0     2015   4/10/2015   
126072      63037  201504110ORL   NBA        1     2015   4/11/2015   
126094      63048  201504130ATL   NBA        1     2015   4/13/2015   
126147      63074  201504150NYK   NBA        0     2015   4/15/2015   

        seasongame  is_playoffs team_id fran_id  pts      elo_i      elo_n  \
1                1            0     NYK  Knicks   68  1300.0000  1306.7233   
3                2            0     NYK  Knicks   47  1306.7233  1297.0712   
23               3            0     NYK  Knicks   68  1297.0712  1307.3197   
37               4            0     NYK  Knicks   68  1307.3197  1293.2161   
54               5            0     NYK  Knicks   64  1293.2161  1296.2185   
...            ...          ...     ...     ...  ...        ...        ...   
126025          78            0     NYK  Knicks   86  1256.8353  1248.3544   
126056          79            0     NYK  Knicks   91  1248.3544  1242.4070   
126072          80            0     NYK  Knicks   80  1242.4070  1249.4366   
126094          81            0     NYK  Knicks  112  1249.4366  1268.4222   
126147          82            0     NYK  Knicks   90  1268.4222  1255.6846   

        win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
1       41.705170    TRH  Huskies       66  1300.0000  1293.2767   
3       40.692783    CHS    Stags       63  1300.0000  1309.6521   
23      41.767715    STB  Bombers       63  1312.8473  1302.5988   
37      40.288467    CHS    Stags       78  1315.2522  1329.3558   
54      40.603352    PIT  Ironmen       62  1277.5587  1274.5563   
...           ...    ...      ...      ...        ...        ...   
126025  17.412123    IND   Pacers      102  1478.9901  1487.4711   
126056  16.998409    MIL    Bucks       99  1452.2738  1458.2212   
126072  17.488194    ORL    Magic       79  1325.8820  1318.8523   
126094  18.861967    ATL    Hawks      108  1618.9574  1599.9718   
126147  17.932079    DET  Pistons      112  1448.3408  1461.0784   

       game_location game_result  forecast notes  
1                  A           W  0.359935   NaN  
3                  A           L  0.368899   NaN  
23                 A           W  0.339290   NaN  
37                 H           L  0.629479   NaN  
54                 H           W  0.660557   NaN  
...              ...         ...       ...   ...  
126025             H           L  0.331108   NaN  
126056             H           L  0.354754   NaN  
126072             A           W  0.258043   NaN  
126094             A           W  0.062810   NaN  
126147             H           L  0.386974   NaN  

[5769 rows x 23 columns]

In [49]:
df.loc[df["fran_id"] == "Knicks","opp_fran"].value_counts()

Celtics         526
Sixers          471
Pistons         382
Hawks           368
Warriors        335
Wizards         323
Lakers          310
Kings           299
Bulls           251
Pacers          205
Bucks           204
Cavaliers       203
Nets            189
Rockets         145
Heat            137
Thunder         124
Clippers        122
Suns            120
Trailblazers    105
Magic            99
Pelicans         90
Jazz             90
Spurs            90
Raptors          85
Nuggets          81
Mavericks        68
Baltimore        62
Timberwolves     50
Hornets          40
Grizzlies        37
Capitols         32
Stags            23
Bombers          23
Olympians        20
Steamrollers     20
Rebels            9
Falcons           6
Ironmen           6
Huskies           6
Jets              5
Packers           2
Denver            2
Redskins          2
Waterloo          2
Name: opp_fran, dtype: int64

In [51]:
df.loc[df["year_id"] == 2010,:]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
111032      55517  200910270CLE   NBA        1     2010  10/27/2009   
111033      55517  200910270CLE   NBA        0     2010  10/27/2009   
111034      55518  200910270DAL   NBA        1     2010  10/27/2009   
111035      55518  200910270DAL   NBA        0     2010  10/27/2009   
111036      55519  200910270LAL   NBA        1     2010  10/27/2009   
...           ...           ...   ...      ...      ...         ...   
113651      56826  201006130BOS   NBA        0     2010   6/13/2010   
113652      56827  201006150LAL   NBA        1     2010   6/15/2010   
113653      56827  201006150LAL   NBA        0     2010   6/15/2010   
113654      56828  201006170LAL   NBA        0     2010   6/17/2010   
113655      56828  201006170LAL   NBA        1     2010   6/17/2010   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
111032           1            0     BOS    Celtics   95  1616.2000  1629.0884   
111033           1            0     CLE  Cavaliers   89  1682.4541  1669.5657   
111034           1            0     WAS    Wizards  102  1361.4187  1387.2084   
111035           1            0     DAL  Mavericks   91  1586.4293  1560.6396   
111036           1            0     LAC   Clippers   92  1317.8915  1317.2549   
...            ...          ...     ...        ...  ...        ...        ...   
113651         104            1     BOS    Celtics   92  1675.2927  1680.6708   
113652         105            1     BOS    Celtics   67  1680.6708  1668.9438   
113653         104            1     LAL     Lakers   89  1679.5972  1691.3241   
113654         105            1     LAL     Lakers   83  1691.3241  1695.1354   
113655         106            1     BOS    Celtics   79  1668.9438  1665.1326   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
111032  53.780228    CLE  Cavaliers       89  1682.4541  1669.5657   
111033  57.480888    BOS    Celtics       95  1616.2000  1629.0884   
111034  29.191736    DAL  Mavericks       91  1586.4293  1560.6396   
111035  46.996815    WAS    Wizards      102  1361.4187  1387.2084   
111036  22.856619    LAL     Lakers       99  1718.7451  1719.3817   
...           ...    ...        ...      ...        ...        ...   
113651  58.446255    LAL     Lakers       86  1684.9752  1679.5972   
113652  57.426167    LAL     Lakers       89  1679.5972  1691.3241   
113653  59.350689    BOS    Celtics       67  1680.6708  1668.9438   
113654  59.668987    BOS    Celtics       79  1668.9438  1665.1326   
113655  57.089245    LAL     Lakers       83  1691.3241  1695.1354   

       game_location game_result  forecast notes  
111032             A           W  0.277472   NaN  
111033             H           L  0.722528   NaN  
111034             A           W  0.133436   NaN  
111035             H           L  0.866564   NaN  
111036             A           L  0.052993   NaN  
...              ...         ...       ...   ...  
113651             H           W  0.627127   NaN  
113652             A           L  0.361360   NaN  
113653             H           W  0.638640   NaN  
113654             H           W  0.669180   NaN  
113655             A           L  0.330820   NaN  

[2624 rows x 23 columns]

In [52]:
df.head(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767   40.29483    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7233   41.70517    TRH   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   
1  Huskies       66     1300.0  1293.2767             A           W  0.359935   

  notes  
0   NaN  
1   NaN

In [55]:
df.loc[(df['year_id'] == 1980)&(df['fran_id'] == 'Knicks'),:]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
41742      20872  197910120ATL   NBA        1     1980  10/12/1979   
41769      20885  197910130NYK   NBA        0     1980  10/13/1979   
41792      20897  197910160NYK   NBA        0     1980  10/16/1979   
41816      20909  197910180DET   NBA        1     1980  10/18/1979   
41830      20916  197910190NJN   NBA        1     1980  10/19/1979   
...          ...           ...   ...      ...      ...         ...   
43444      21723  198003220NYK   NBA        0     1980   3/22/1980   
43457      21729  198003230PHI   NBA        1     1980   3/23/1980   
43468      21735  198003250CLE   NBA        1     1980   3/25/1980   
43481      21741  198003260BOS   NBA        1     1980   3/26/1980   
43500      21751  198003270NYK   NBA        0     1980   3/27/1980   

       seasongame  is_playoffs team_id fran_id  pts      elo_i      elo_n  \
41742           1            0     NYK  Knicks  104  1417.7286  1412.8352   
41769           2            0     NYK  Knicks  125  1412.8352  1421.9812   
41792           3            0     NYK  Knicks  126  1421.9812  1428.8911   
41816           4            0     NYK  Knicks  115  1428.8911  1421.2010   
41830           5            0     NYK  Knicks  111  1421.2010  1430.3867   
...           ...          ...     ...     ...  ...        ...        ...   
43444          78            0     NYK  Knicks  113  1462.0233  1447.9253   
43457          79            0     NYK  Knicks  108  1447.9253  1445.9675   
43468          80            0     NYK  Knicks  128  1445.9675  1466.1354   
43481          81            0     NYK  Knicks  121  1466.1354  1463.9615   
43500          82            0     NYK  Knicks  101  1463.9615  1459.9078   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
41742  31.612125    ATL      Hawks      121  1559.2893  1564.1827   
41769  32.525932    WSB    Wizards      118  1536.1177  1526.9717   
41792  33.222260    HOU    Rockets      121  1517.0374  1510.1274   
41816  32.447624    DET    Pistons      129  1452.9895  1460.6796   
41830  33.373604    NJN       Nets      106  1399.9999  1390.8142   
...          ...    ...        ...      ...        ...        ...   
43444  35.163406    WSB    Wizards      122  1448.1752  1462.2732   
43457  34.962353    PHI     Sixers      112  1603.8334  1605.7911   
43468  37.045979    CLE  Cavaliers      115  1507.2769  1487.1090   
43481  36.820198    BOS    Celtics      129  1669.1906  1671.3644   
43500  36.399860    PHI     Sixers      103  1609.5419  1613.5956   

      game_location game_result  forecast notes  
41742             A           L  0.199322   NaN  
41769             H           W  0.466544   NaN  
41792             H           W  0.507114   NaN  
41816             A           L  0.328634   NaN  
41830             A           W  0.388505   NaN  
...             ...         ...       ...   ...  
43444             H           L  0.658218   NaN  
43457             A           L  0.186468   NaN  
43468             A           W  0.283215   NaN  
43481             A           L  0.148739   NaN  
43500             H           L  0.434778   NaN  

[82 rows x 23 columns]

### Kombinasi .loc dan fungsi agregasi

In [56]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

1947

team dengan fran_id huskies bermain maksinal hingga season 1947 saja, atau hanya bermain 1 season di data record

In [64]:
df.loc[df["team_id"] == "NYK","year_id"].max()

2015

In [67]:
df.loc[df["team_id"] == "NYK","year_id"].agg(["min","max"])

min    1947
max    2015
Name: year_id, dtype: int64

team dengan fran_id Knick bermain selama season 1947 hingga 2015

In [71]:
# df.sample()
# nge return data random
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
27664      13833  197301090UTS   ABA        0     1973    1/9/1973   
38798      19400  197801140BUF   NBA        1     1978   1/14/1978   
62124      31063  198911170UTA   NBA        0     1990  11/17/1989   
6635        3318  195511130FTW   NBA        0     1956  11/13/1955   
51684      25843  198410300UTA   NBA        0     1985  10/30/1984   

       seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
27664          45            0     UTS    Stars  115  1601.4862  1602.7350   
38798          40            0     WSB  Wizards   93  1588.2084  1554.6177   
62124           7            0     UTA     Jazz  114  1542.3317  1550.2230   
6635            4            0     FTW  Pistons   93  1527.9282  1539.5039   
51684           3            0     UTA     Jazz  117  1471.5483  1479.1093   

       win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
27664  55.055866    MMT    Sounds      111  1364.0497  1362.8009   
38798  46.369865    BUF  Clippers      130  1406.4707  1440.0614   
62124  48.510319    IND    Pacers      100  1523.8827  1515.9913   
6635   42.296356    SYR    Sixers       79  1600.0917  1588.5160   
51684  38.607769    NYK    Knicks      111  1565.4785  1557.9176   

      game_location game_result  forecast notes  
27664             H           W  0.874620   NaN  
38798             A           L  0.615507   NaN  
62124             H           W  0.664151   NaN  
6635              H           W  0.539975   NaN  
51684             H           W  0.508734   NaN

In [72]:
df.loc[df['fran_id'] == 'Warriors', 'pts'].sum()

591224

In [74]:
df.loc[df['fran_id'] == 'Warriors', ['seasongame','pts']].sum()

seasongame    237531
pts           591224
dtype: int64

### Menggunakan .iloc

In [76]:
df.sample(2)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
18237        9119  196901180INA   ABA        1     1969   1/18/1969   
124009      62005  201411190BRK   NBA        0     2015  11/19/2014   

        seasongame  is_playoffs team_id fran_id  pts     elo_i      elo_n  \
18237           41            0     LAS   Stars  129  1301.429  1308.4583   
124009          11            0     BRK    Nets  118  1501.950  1489.5875   

        win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
18237   36.529182    INA   Pacers      128  1384.8793  1377.8500   
124009  39.432415    MIL    Bucks      122  1367.5092  1379.8716   

       game_location game_result  forecast notes  
18237              A           W  0.258070   NaN  
124009             H           L  0.794056   NaN

In [89]:
df.iloc[18237:19000,:]

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
18237       9119  196901180INA   ABA        1     1969  1/18/1969          41   
18238       9120  196901180KEN   ABA        1     1969  1/18/1969          40   
18239       9120  196901180KEN   ABA        0     1969  1/18/1969          41   
18240       9121  196901180MNP   ABA        0     1969  1/18/1969          41   
18241       9121  196901180MNP   ABA        1     1969  1/18/1969          38   
...          ...           ...   ...      ...      ...        ...         ...   
18995       9498  196903160NYA   ABA        0     1969  3/16/1969          70   
18996       9499  196903170PHO   NBA        0     1969  3/17/1969          79   
18997       9499  196903170PHO   NBA        1     1969  3/17/1969          80   
18998       9500  196903170LAS   ABA        1     1969  3/17/1969          70   
18999       9500  196903170LAS   ABA        0     1969  3/17/1969          69   

       is_playoffs team_id   fran_id  pts      elo_i      elo_n  win_equiv  \
18237            0     LAS     Stars  129  1301.4290  1308.4583  36.529182   
18238            0     HSM   Spirits  102  1252.1792  1247.9250  30.407274   
18239            0     KEN  Colonels  112  1354.2355  1358.4896  41.759319   
18240            0     MNP   Condors  108  1401.4327  1396.6979  45.745911   
18241            0     OAK   Squires  111  1544.6616  1549.3965  60.169971   
...            ...     ...       ...  ...        ...        ...        ...   
18995            0     NYA      Nets  103  1159.2981  1151.8845  21.730261   
18996            0     PHO      Suns   95  1207.2982  1189.9410  17.571697   
18997            0     DET   Pistons  119  1326.0577  1343.4149  30.689789   
18998            0     NOB    Sounds  118  1449.6981  1461.8121  52.297749   
18999            0     LAS     Stars   99  1282.0728  1269.9587  32.593765   

      opp_id    opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location  \
18237    INA      Pacers      128  1384.8793  1377.8500             A   
18238    KEN    Colonels      112  1354.2355  1358.4896             A   
18239    HSM     Spirits      102  1252.1792  1247.9250             H   
18240    OAK     Squires      111  1544.6616  1549.3965             H   
18241    MNP     Condors      108  1401.4327  1396.6979             A   
...      ...         ...      ...        ...        ...           ...   
18995    MMF  Floridians      116  1380.8054  1388.2190             H   
18996    DET     Pistons      119  1326.0577  1343.4149             H   
18997    PHO        Suns       95  1207.2982  1189.9410             A   
18998    LAS       Stars       99  1282.0728  1269.9587             A   
18999    NOB      Sounds      118  1449.6981  1461.8121             H   

      game_result  forecast          notes  
18237           W  0.258070            NaN  
18238           L  0.238099            NaN  
18239           W  0.761901            NaN  
18240           L  0.438108            NaN  
18241           W  0.561892            NaN  
...           ...       ...            ...  
18995           L  0.331934            NaN  
18996           L  0.473029            NaN  
18997           W  0.526971            NaN  
18998           W  0.596110  at Jackson MS  
18999           L  0.403890  at Jackson MS  

[763 rows x 23 columns]

In [91]:
df.loc[18237:19000,['game_id','year_id']]

game_id  year_id
18237  196901180INA     1969
18238  196901180KEN     1969
18239  196901180KEN     1969
18240  196901180MNP     1969
18241  196901180MNP     1969
...             ...      ...
18996  196903170PHO     1969
18997  196903170PHO     1969
18998  196903170LAS     1969
18999  196903170LAS     1969
19000  196903180BAL     1969

[764 rows x 2 columns]

## Query dataset

In [93]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
67508       33755  199201090CHH   NBA        0     1992    1/9/1992   
102925      51463  200604170MIL   NBA        1     2006   4/17/2006   
62054       31028  198911120MIN   NBA        0     1990  11/12/1989   
105536      52769  200704140UTA   NBA        1     2007   4/14/2007   
71642       35822  199311260DEN   NBA        1     1994  11/26/1993   

        seasongame  is_playoffs team_id       fran_id  pts      elo_i  \
67508           34            0     CHH      Pelicans  109  1359.0313   
102925          81            0     DET       Pistons   93  1704.1940   
62054            5            0     MIN  Timberwolves   97  1301.8452   
105536          80            0     PHO          Suns  126  1672.3174   
71642           11            0     POR  Trailblazers  101  1555.6187   

            elo_n  win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
67508   1366.5684  28.494591    SAC    Kings       96  1341.1473  1333.6102   
102925  1678.3438  58.606575    MIL    Bucks      113  1459.0922  1484.9424   
62054   1294.9655  23.107161    SEA  Thunder      108  1516.3013  1523.1810   
105536  1692.5206  59.672935    UTA     Jazz       98  1564.4889  1544.2856   
71642   1544.4585  46.016273    DEN  Nuggets      112  1451.9569  1463.1171   

       game_location game_result  forecast notes  
67508              H           W  0.663425   NaN  
102925             A           L  0.697469   NaN  
62054              H           L  0.340995   NaN  
105536             A           W  0.511264   NaN  
71642              A           L  0.505269   NaN

In [94]:
#single
df[
    df['year_id']==2001
]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87750      43876  200010310ATL   NBA        0     2001  10/31/2000   
87751      43876  200010310ATL   NBA        1     2001  10/31/2000   
87752      43877  200010310CHI   NBA        0     2001  10/31/2000   
87753      43877  200010310CHI   NBA        1     2001  10/31/2000   
87754      43878  200010310DAL   NBA        0     2001  10/31/2000   
...          ...           ...   ...      ...      ...         ...   
90265      45133  200106100PHI   NBA        0     2001   6/10/2001   
90266      45134  200106130PHI   NBA        0     2001   6/13/2001   
90267      45134  200106130PHI   NBA        1     2001   6/13/2001   
90268      45135  200106150PHI   NBA        0     2001   6/15/2001   
90269      45135  200106150PHI   NBA        1     2001   6/15/2001   

       seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
87750           1            0     ATL      Hawks   82  1366.9685  1354.0496   
87751           1            0     CHH   Pelicans  106  1557.8643  1570.7832   
87752           1            0     CHI      Bulls   81  1326.7501  1316.6208   
87753           1            0     SAC      Kings  100  1535.6016  1545.7308   
87754           1            0     DAL  Mavericks   97  1541.3955  1545.2712   
...           ...          ...     ...        ...  ...        ...        ...   
90265         103            1     PHI     Sixers   91  1605.4132  1599.5922   
90266         104            1     PHI     Sixers   86  1599.5922  1589.4489   
90267          97            1     LAL     Lakers  100  1760.3232  1770.4666   
90268         105            1     PHI     Sixers   96  1589.4489  1581.0239   
90269          98            1     LAL     Lakers  108  1770.4666  1778.8915   

       win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87750  26.311760    CHH  Pelicans      106  1557.8643  1570.7832   
87751  48.286263    ATL     Hawks       82  1366.9685  1354.0496   
87752  23.012793    SAC     Kings      100  1535.6016  1545.7308   
87753  45.715614    CHI     Bulls       81  1326.7501  1316.6208   
87754  45.668037    MIL     Bucks       93  1522.6825  1518.8068   
...          ...    ...       ...      ...        ...        ...   
90265  51.169765    LAL    Lakers       96  1754.5022  1760.3232   
90266  50.164997    LAL    Lakers      100  1760.3232  1770.4666   
90267  65.530334    PHI    Sixers       86  1599.5922  1589.4489   
90268  49.321487    LAL    Lakers      108  1770.4666  1778.8915   
90269  66.090416    PHI    Sixers       96  1589.4489  1581.0239   

      game_location game_result  forecast notes  
87750             H           L  0.372095   NaN  
87751             A           W  0.627905   NaN  
87752             H           L  0.348282   NaN  
87753             A           W  0.651718   NaN  
87754             H           W  0.664490   NaN  
...             ...         ...       ...   ...  
90265             H           L  0.429822   NaN  
90266             H           L  0.413480   NaN  
90267             A           W  0.586520   NaN  
90268             H           L  0.385474   NaN  
90269             A           W  0.614525   NaN  

[2520 rows x 23 columns]

In [ ]:
#Multiple

df[
    (df['year_id']==2001) &
    (df['fran_id']== 'Celtics') &
    (df['game_location']=='H') &
    (df['game_result']== 'W')
] [['fran_id','pts','opp_fran','opp_pts']]

In [103]:
df.query("year_id == 2001 & fran_id == 'Celtics' & game_location == 'H' & game_result == 'W' ")

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87776      43889  200011010BOS   NBA        0     2001   11/1/2000   
87805      43903  200011030BOS   NBA        0     2001   11/3/2000   
87970      43986  200011150BOS   NBA        0     2001  11/15/2000   
88001      44001  200011170BOS   NBA        0     2001  11/17/2000   
88068      44035  200011220BOS   NBA        0     2001  11/22/2000   
88093      44047  200011240BOS   NBA        0     2001  11/24/2000   
88233      44117  200012030BOS   NBA        0     2001   12/3/2000   
88336      44169  200012100BOS   NBA        0     2001  12/10/2000   
88469      44235  200012200BOS   NBA        0     2001  12/20/2000   
88748      44375  200101100BOS   NBA        0     2001   1/10/2001   
88925      44463  200101230BOS   NBA        0     2001   1/23/2001   
88975      44488  200101260BOS   NBA        0     2001   1/26/2001   
89013      44507  200101290BOS   NBA        0     2001   1/29/2001   
89105      44553  200102040BOS   NBA        0     2001    2/4/2001   
89141      44571  200102070BOS   NBA        0     2001    2/7/2001   
89364      44683  200102260BOS   NBA        0     2001   2/26/2001   
89425      44713  200103020BOS   NBA        0     2001    3/2/2001   
89525      44763  200103090BOS   NBA        0     2001    3/9/2001   
89756      44879  200103250BOS   NBA        0     2001   3/25/2001   
90057      45029  200104150BOS   NBA        0     2001   4/15/2001   

       seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
87776           1            0     BOS  Celtics  103  1496.9196  1509.4957   
87805           2            0     BOS  Celtics   93  1509.4957  1512.2125   
87970           7            0     BOS  Celtics  116  1483.8416  1487.9716   
88001           8            0     BOS  Celtics  106  1487.9716  1490.5791   
88068          11            0     BOS  Celtics   96  1460.4922  1472.5193   
88093          12            0     BOS  Celtics  103  1472.5193  1477.3024   
88233          17            0     BOS  Celtics   96  1464.1198  1466.8375   
88336          20            0     BOS  Celtics  104  1450.2649  1453.4750   
88469          25            0     BOS  Celtics   99  1397.2222  1405.6613   
88748          36            0     BOS  Celtics   88  1368.3331  1384.8729   
88925          42            0     BOS  Celtics   83  1384.1893  1386.8386   
88975          43            0     BOS  Celtics   99  1386.8386  1404.1323   
89013          44            0     BOS  Celtics   99  1404.1323  1412.0132   
89105          47            0     BOS  Celtics  103  1437.8401  1448.2347   
89141          49            0     BOS  Celtics   94  1442.6997  1449.4445   
89364          57            0     BOS  Celtics   85  1468.3358  1472.3491   
89425          59            0     BOS  Celtics  103  1468.3030  1476.5731   
89525          62            0     BOS  Celtics  104  1451.1699  1459.3822   
89756          69            0     BOS  Celtics  102  1477.7058  1479.0504   
90057          81            0     BOS  Celtics   95  1463.9949  1467.3157   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87776  41.935509    DET    Pistons       83  1529.7430  1517.1669   
87805  42.220291    TOR    Raptors       91  1472.1443  1469.4275   
87970  39.678341    WAS    Wizards      109  1418.0574  1413.9274   
88001  39.951633    ATL      Hawks       99  1332.2992  1329.6917   
88068  38.062054    HOU    Rockets       81  1530.7407  1518.7136   
88093  38.561615    ORL      Magic       98  1476.9380  1472.1549   
88233  37.469780    LAC   Clippers       87  1289.0637  1286.3459   
88336  36.083122    DEN    Nuggets      102  1449.5248  1446.3147   
88469  31.235086    NJN       Nets       86  1404.6704  1396.2313   
88748  29.205793    MIA       Heat       76  1575.6382  1559.0984   
88925  29.395197    ATL      Hawks       82  1380.4136  1377.7643   
88975  31.083946    PHO       Suns       83  1546.8582  1529.5645   


In [105]:
df.query(
    "year_id == 2001 &\
    fran_id == 'Celtics' &\
    game_location == 'H' &\
    game_result == 'W' ")

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87776      43889  200011010BOS   NBA        0     2001   11/1/2000   
87805      43903  200011030BOS   NBA        0     2001   11/3/2000   
87970      43986  200011150BOS   NBA        0     2001  11/15/2000   
88001      44001  200011170BOS   NBA        0     2001  11/17/2000   
88068      44035  200011220BOS   NBA        0     2001  11/22/2000   
88093      44047  200011240BOS   NBA        0     2001  11/24/2000   
88233      44117  200012030BOS   NBA        0     2001   12/3/2000   
88336      44169  200012100BOS   NBA        0     2001  12/10/2000   
88469      44235  200012200BOS   NBA        0     2001  12/20/2000   
88748      44375  200101100BOS   NBA        0     2001   1/10/2001   
88925      44463  200101230BOS   NBA        0     2001   1/23/2001   
88975      44488  200101260BOS   NBA        0     2001   1/26/2001   
89013      44507  200101290BOS   NBA        0     2001   1/29/2001   
89105      44553  200102040BOS   NBA        0     2001    2/4/2001   
89141      44571  200102070BOS   NBA        0     2001    2/7/2001   
89364      44683  200102260BOS   NBA        0     2001   2/26/2001   
89425      44713  200103020BOS   NBA        0     2001    3/2/2001   
89525      44763  200103090BOS   NBA        0     2001    3/9/2001   
89756      44879  200103250BOS   NBA        0     2001   3/25/2001   
90057      45029  200104150BOS   NBA        0     2001   4/15/2001   

       seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
87776           1            0     BOS  Celtics  103  1496.9196  1509.4957   
87805           2            0     BOS  Celtics   93  1509.4957  1512.2125   
87970           7            0     BOS  Celtics  116  1483.8416  1487.9716   
88001           8            0     BOS  Celtics  106  1487.9716  1490.5791   
88068          11            0     BOS  Celtics   96  1460.4922  1472.5193   
88093          12            0     BOS  Celtics  103  1472.5193  1477.3024   
88233          17            0     BOS  Celtics   96  1464.1198  1466.8375   
88336          20            0     BOS  Celtics  104  1450.2649  1453.4750   
88469          25            0     BOS  Celtics   99  1397.2222  1405.6613   
88748          36            0     BOS  Celtics   88  1368.3331  1384.8729   
88925          42            0     BOS  Celtics   83  1384.1893  1386.8386   
88975          43            0     BOS  Celtics   99  1386.8386  1404.1323   
89013          44            0     BOS  Celtics   99  1404.1323  1412.0132   
89105          47            0     BOS  Celtics  103  1437.8401  1448.2347   
89141          49            0     BOS  Celtics   94  1442.6997  1449.4445   
89364          57            0     BOS  Celtics   85  1468.3358  1472.3491   
89425          59            0     BOS  Celtics  103  1468.3030  1476.5731   
89525          62            0     BOS  Celtics  104  1451.1699  1459.3822   
89756          69            0     BOS  Celtics  102  1477.7058  1479.0504   
90057          81            0     BOS  Celtics   95  1463.9949  1467.3157   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87776  41.935509    DET    Pistons       83  1529.7430  1517.1669   
87805  42.220291    TOR    Raptors       91  1472.1443  1469.4275   
87970  39.678341    WAS    Wizards      109  1418.0574  1413.9274   
88001  39.951633    ATL      Hawks       99  1332.2992  1329.6917   
88068  38.062054    HOU    Rockets       81  1530.7407  1518.7136   
88093  38.561615    ORL      Magic       98  1476.9380  1472.1549   
88233  37.469780    LAC   Clippers       87  1289.0637  1286.3459   
88336  36.083122    DEN    Nuggets      102  1449.5248  1446.3147   
88469  31.235086    NJN       Nets       86  1404.6704  1396.2313   
88748  29.205793    MIA       Heat       76  1575.6382  1559.0984   
88925  29.395197    ATL      Hawks       82  1380.4136  1377.7643   
88975  31.083946    PHO       Suns       83  1546.8582  1529.5645   


In [162]:
df.query(
    "year_id == 2001 &\
    fran_id == 'Celtics' &\
    game_location == 'H' &\
    result == 'W' ")\
    [['fran_id','pts','opp_fran','opp_pts']]

fran_id  pts   opp_fran  opp_pts
87776  Celtics  103    Pistons       83
87805  Celtics   93    Raptors       91
87970  Celtics  116    Wizards      109
88001  Celtics  106      Hawks       99
88068  Celtics   96    Rockets       81
88093  Celtics  103      Magic       98
88233  Celtics   96   Clippers       87
88336  Celtics  104    Nuggets      102
88469  Celtics   99       Nets       86
88748  Celtics   88       Heat       76
88925  Celtics   83      Hawks       82
88975  Celtics   99       Suns       83
89013  Celtics   99    Pistons       87
89105  Celtics  103  Cavaliers       82
89141  Celtics   94      Bucks       91
89364  Celtics   85    Thunder       82
89425  Celtics  103       Jazz       99
89525  Celtics  104    Wizards       72
89756  Celtics  102      Bulls       98
90057  Celtics   95       Nets       88

### Grouping and Aggregating

In [108]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
32282       16142  197501020SAA   ABA        0     1975   1/2/1975   
106692      53347  200801020MIA   NBA        0     2008   1/2/2008   
88825       44413  200101150GSW   NBA        1     2001  1/15/2001   
11711        5856  196302080LAL   NBA        1     1963   2/8/1963   
89789       44895  200103270LAC   NBA        1     2001  3/27/2001   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
32282           40            0     SAA      Spurs  122  1549.1060  1554.3804   
106692          31            0     MIA       Heat   98  1395.2227  1384.7980   
88825           36            0     CLE  Cavaliers  107  1441.4468  1449.9563   
11711           61            0     CHZ    Wizards   91  1341.2123  1340.4452   
89789           71            0     DET    Pistons   94  1397.0587  1392.2009   

        win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
32282   48.565411    SDA     Sails      104  1414.3739  1409.0995   
106692  29.105276    MIL     Bucks      103  1373.3943  1383.8190   
88825   35.719738    GSW  Warriors      101  1368.2139  1359.7043   
11711   25.986189    LAL    Lakers       94  1641.8629  1642.6300   
89789   29.914600    LAC  Clippers      101  1428.7046  1433.5624   

       game_location game_result  forecast notes  
32282              H           W  0.794330   NaN  
106692             H           L  0.668476   NaN  
88825              A           W  0.461555   NaN  
11711              A           L  0.090600   NaN  
89789              A           L  0.319121   NaN

In [109]:
#mengelompokan data

gr = df.groupby('game_result')
gr

In [110]:
#simulasi saja untuk menampilkan

gr.get_group('L')

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
6               4  194611020PRO   NBA        1     1947  11/2/1946   
8               5  194611020STB   NBA        1     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126305      63153  201506070GSW   NBA        0     2015   6/7/2015   
126306      63154  201506090CLE   NBA        1     2015   6/9/2015   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
0                1            0     TRH    Huskies   66  1300.0000  1293.2767   
3                2            0     NYK     Knicks   47  1306.7233  1297.0712   
4                1            0     DTF    Falcons   33  1300.0000  1279.6189   
6                1            0     BOS    Celtics   53  1300.0000  1294.8458   
8                1            0     PIT    Ironmen   51  1300.0000  1295.3092   
...            ...          ...     ...        ...  ...        ...        ...   
126305          99            1     GSW   Warriors   93  1806.2035  1797.5032   
126306         100            1     GSW   Warriors   91  1797.5032  1790.9591   
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   

        win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
0       40.294830    NYK        Knicks       68  1300.0000  1306.7233   
3       40.692783    CHS         Stags       63  1300.0000  1309.6521   
4       38.864048    WSC      Capitols       50  1300.0000  1320.3811   
6       40.459381    PRO  Steamrollers       59  1300.0000  1305.1542   
8       40.507980    STB       Bombers       56  1300.0000  1304.6908   
...           ...    ...           ...      ...        ...        ...   
126305  67.029549    CLE     Cavaliers       95  1708.1705  1716.8708   
126306  66.615417    CLE     Cavaliers       96  1716.8708  1723.4149   
126309  60.309792    GSW      Warriors      103  1790.9591  1809.9791   
126311  60.010067    GSW      Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW      Warriors      105  1813.6349  1822.2881   

       game_location game_result  forecast notes  
0                  H           L  0.640065   NaN  
3                  A           L  0.368899   NaN  
4                  H           L  0.640065   NaN  
6                  A           L  0.359935   NaN  
8                  A           L  0.359935   NaN  
...              ...         ...       ...   ...  
126305             H           L  0.757674   NaN  
126306             A           L  0.472157   NaN  
126309             H           L  0.546572   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  

[63157 rows x 23 columns]

In [119]:
#aggregating
df.groupby('game_result')['pts'].mean()

#round(df.groupby('game_result')['pts'].mean(),2)
#jika ingin dituliskan dengan desimal tidak lebih dari

game_result
L     97.349209
W    108.110756
Name: pts, dtype: float64

rata-rata pts untuk group team yang kalah(L) adalah 97.34, sedangkan untuk rata-rata pts group yang menang adalah 108.11

In [165]:
#kombinasi group by dan query
# df.groupby('A').agg({'B': ['min', 'max'], 'C': 'sum'})

df[
    (df['year_id']>2010)&
    (df['fran_id']=='Spurs')
].groupby(['year_id','result']).agg({'game_id':'count'})


game_id
year_id result         
2011    L            25
        W            63
2012    L            20
        W            60
2013    L            30
        W            73
2014    L            27
        W            78
2015    L            31
        W            58

## Manipulating Colums
- create new columns base on existing data
- rename colums
- drop colums

In [136]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
26334       13168  197203280FLO   ABA        1     1972   3/28/1972   
81945       40973  199712300DET   NBA        1     1998  12/30/1997   
82642       41322  199802200NJN   NBA        1     1998   2/20/1998   
51540       25771  198404260POR   NBA        0     1984   4/26/1984   
103027      51514  200604300CHI   NBA        0     2006   4/30/2006   

        seasongame  is_playoffs team_id       fran_id  pts      elo_i  \
26334           83            0     MMP        Sounds  107  1279.9531   
81945           29            0     TOR       Raptors   95  1339.3318   
82642           53            0     CLE     Cavaliers  109  1512.4641   
51540           87            1     POR  Trailblazers  105  1523.6027   
103027          86            1     CHI         Bulls   93  1549.9109   

            elo_n  win_equiv opp_id    opp_fran  opp_pts  opp_elo_i  \
26334   1275.6639  24.698513    FLO  Floridians      118  1392.0222   
81945   1337.3309  25.347132    DET     Pistons      100  1511.0435   
82642   1533.0145  45.025646    NJN        Nets       95  1562.2736   
51540   1507.9834  41.740692    PHO        Suns      117  1535.6576   
103027  1555.3651  46.886177    MIA        Heat       87  1562.8293   

        opp_elo_n game_location game_result  forecast notes  
26334   1396.3114             A           L  0.227801   NaN  
81945   1513.0443             A           L  0.173059   NaN  
82642   1541.7231             A           W  0.296844   NaN  
51540   1551.2769             H           L  0.623928   NaN  
103027  1557.3751             H           W  0.622760   NaN

In [147]:
# create new columns based on existing data

df['selisih poin'] = abs(df['pts'] - df['opp_pts'])


In [148]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  selisih poin  
0  0.640065   NaN             2  
1  0.359935   NaN             2  
2  0.631101   NaN            16  
3  0.368899   NaN            16  
4  0.640065   NaN            17

In [152]:
#rename colums
df.rename(columns={"game_result": "result"},inplace=True)

#rename tidak mensave hasil perubahan, untuk men save perubahan maka
#df.rename(columns={"game_result": "result"},inplace=True)

In [154]:
df.head(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location result  forecast  \
0    Knicks       68  1300.0000  1306.7233             H      L  0.640065   
1   Huskies       66  1300.0000  1293.2767             A      W  0.359935   
2    Knicks       47  1306.7233  1297.0712             H      W  0.631101   
3     Stags       63  1300.0000  1309.6521             A      L  0.368899   
4  Capitols       50  1300.0000  1320.3811             H      L  0.640065   

  notes  selisih poin  
0   NaN             2  
1   NaN             2  
2   NaN            16  
3   NaN            16  
4   NaN            17

In [ ]:
# drop columns
df.drop(['elo_i', 'elo_n'], axis =1, inplace=True)

In [160]:
df

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts  win_equiv opp_id  \
0                1            0     TRH    Huskies   66  40.294830    NYK   
1                1            0     NYK     Knicks   68  41.705170    TRH   
2                1            0     CHS      Stags   63  42.012257    NYK   
3                2            0     NYK     Knicks   47  40.692783    CHS   
4                1            0     DTF    Falcons   33  38.864048    WSC   
...            ...          ...     ...        ...  ...        ...    ...   
126309         100            1     CLE  Cavaliers   82  60.309792    GSW   
126310         102            1     GSW   Warriors  104  68.013329    CLE   
126311         101            1     CLE  Cavaliers   91  60.010067    GSW   
126312         102            1     CLE  Cavaliers   97  59.290245    GSW   
126313         103            1     GSW   Warriors  105  68.519516    CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location result  \
0          Knicks       68  1300.0000  1306.7233             H      L   
1         Huskies       66  1300.0000  1293.2767             A      W   
2          Knicks       47  1306.7233  1297.0712             H      W   
3           Stags       63  1300.0000  1309.6521             A      L   
4        Capitols       50  1300.0000  1320.3811             H      L   
...           ...      ...        ...        ...           ...    ...   
126309   Warriors      103  1790.9591  1809.9791             H      L   
126310  Cavaliers       91  1704.3949  1700.7391             H      W   
126311   Warriors      104  1809.9791  1813.6349             A      L   
126312   Warriors      105  1813.6349  1822.2881             H      L   
126313  Cavaliers       97  1700.7391  1692.0859             A      W   

        forecast notes  selisih poin  
0       0.640065   NaN             2  
1       0.359935   NaN             2  
2       0.631101   NaN            16  
3       0.368899   NaN            16  
4       0.640065   NaN            17  
...          ...   ...           ...  
126309  0.546572   NaN            21  
126310  0.765565   NaN            13  
126311  0.234435   NaN            13  
126312  0.481450   NaN             8  
126313  0.518550   NaN             8  

[126314 rows x 22 columns]